core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
> Do we make it too complicated by allow effects to influence the construction of metadata, or do we make it too complicated by forcing effects into keys.

But if effects influence the construction of the map, you STILL have to have effects in the keys. There can be arbitrarily many kinds of metadata, with widely varying expensiveness, importance, locality, etc. Choosing between all of it and none of it isn't much of a choice.

I'd far prefer a second method like partial functions have - `isDefinedAt`, or `hasMetadataFor`, or whatever - and have the lookup method return a map.